File: implementations/go/dijkstra.go
package main

import (
	"container/heap"
	"fmt"
	"math"
)

// Edge is one directed, weighted connection in a trusted graph.
type Edge struct {
	To     string
	Weight float64
}

type Graph map[string][]Edge

type Result struct {
	Found bool
	Cost  *float64
	Path  []string
}

type queueEntry struct {
	node     string
	distance float64
}

type minHeap []queueEntry

func (h minHeap) Len() int { return len(h) }
func (h minHeap) Less(i, j int) bool {
	return h[i].distance < h[j].distance ||
		(h[i].distance == h[j].distance && h[i].node < h[j].node)
}
func (h minHeap) Swap(i, j int) { h[i], h[j] = h[j], h[i] }
func (h *minHeap) Push(value any) { *h = append(*h, value.(queueEntry)) }
func (h *minHeap) Pop() any {
	old := *h
	last := old[len(old)-1]
	*h = old[:len(old)-1]
	return last
}

func validateGraph(graph Graph, start, goal string) error {
	if _, ok := graph[start]; !ok {
		return fmt.Errorf("start node %q is missing", start)
	}
	if _, ok := graph[goal]; !ok {
		return fmt.Errorf("goal node %q is missing", goal)
	}
	for _, edges := range graph {
		for _, edge := range edges {
			if _, ok := graph[edge.To]; !ok {
				return fmt.Errorf("unknown node %q", edge.To)
			}
			if math.IsNaN(edge.Weight) || math.IsInf(edge.Weight, 0) || edge.Weight < 0 {
				return fmt.Errorf("Dijkstra requires finite, nonnegative weights")
			}
		}
	}
	return nil
}

func dijkstra(graph Graph, start, goal string) (Result, error) {
	if err := validateGraph(graph, start, goal); err != nil {
		return Result{}, err
	}

	distances := make(map[string]float64, len(graph))
	parents := make(map[string]string, len(graph))
	for node := range graph {
		distances[node] = math.Inf(1)
	}
	distances[start] = 0

	frontier := &minHeap{{node: start, distance: 0}}
	heap.Init(frontier)
	for frontier.Len() > 0 {
		current := heap.Pop(frontier).(queueEntry)
		if current.distance != distances[current.node] {
			continue
		}
		if current.node == goal {
			break
		}

		for _, edge := range graph[current.node] {
			candidate := current.distance + edge.Weight
			if candidate >= distances[edge.To] {
				continue
			}

			distances[edge.To] = candidate
			parents[edge.To] = current.node
			heap.Push(frontier, queueEntry{node: edge.To, distance: candidate})
		}
	}

	if math.IsInf(distances[goal], 1) {
		return Result{Found: false, Cost: nil, Path: []string{}}, nil
	}

	path := []string{goal}
	for path[len(path)-1] != start {
		path = append(path, parents[path[len(path)-1]])
	}
	for left, right := 0, len(path)-1; left < right; left, right = left+1, right-1 {
		path[left], path[right] = path[right], path[left]
	}
	cost := distances[goal]
	return Result{Found: true, Cost: &cost, Path: path}, nil
}

func main() {
	graph := Graph{
		"A": {{To: "B", Weight: 2}, {To: "C", Weight: 7}},
		"B": {{To: "C", Weight: 1}, {To: "D", Weight: 5}},
		"C": {{To: "D", Weight: 1}},
		"D": {},
	}
	result, err := dijkstra(graph, "A", "D")
	if err != nil || !result.Found || result.Cost == nil || *result.Cost != 4 ||
		fmt.Sprint(result.Path) != "[A B C D]" {
		panic(fmt.Sprintf("unexpected result: %#v (%v)", result, err))
	}
	fmt.Println("go dijkstra: ok")
}

File: implementations/rust/dijkstra.rs
use std::cmp::Ordering;
use std::collections::{BinaryHeap, HashMap};

// Trusted, repository-controlled Dijkstra reference implementation.
type NodeId = &'static str;
type Graph = HashMap<NodeId, Vec<Edge>>;

#[derive(Clone, Copy)]
struct Edge {
    to: NodeId,
    weight: u64,
}

#[derive(Debug, PartialEq)]
struct DijkstraResult {
    found: bool,
    cost: Option<u64>,
    path: Vec<NodeId>,
}

#[derive(Eq, PartialEq)]
struct QueueEntry {
    node: NodeId,
    distance: u64,
}

impl Ord for QueueEntry {
    fn cmp(&self, other: &Self) -> Ordering {
        other
            .distance
            .cmp(&self.distance)
            .then_with(|| other.node.cmp(self.node))
    }
}

impl PartialOrd for QueueEntry {
    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
        Some(self.cmp(other))
    }
}

fn validate_graph(graph: &Graph, start: NodeId, goal: NodeId) -> Result<(), String> {
    if !graph.contains_key(start) || !graph.contains_key(goal) {
        return Err("start and goal must be present in the graph".into());
    }
    for edges in graph.values() {
        for edge in edges {
            if !graph.contains_key(edge.to) {
                return Err(format!("unknown node: {}", edge.to));
            }
        }
    }
    Ok(())
}

fn dijkstra(graph: &Graph, start: NodeId, goal: NodeId) -> Result<DijkstraResult, String> {
    validate_graph(graph, start, goal)?;

    let mut distances: HashMap<NodeId, u64> =
        graph.keys().map(|&node| (node, u64::MAX)).collect();
    let mut parents: HashMap<NodeId, NodeId> = HashMap::new();
    distances.insert(start, 0);

    let mut frontier = BinaryHeap::new();
    frontier.push(QueueEntry {
        node: start,
        distance: 0,
    });

    while let Some(current) = frontier.pop() {
        if current.distance != distances[current.node] {
            continue;
        }
        if current.node == goal {
            break;
        }

        for edge in &graph[current.node] {
            let Some(candidate) = current.distance.checked_add(edge.weight) else {
                continue;
            };
            if candidate >= distances[edge.to] {
                continue;
            }

            distances.insert(edge.to, candidate);
            parents.insert(edge.to, current.node);
            frontier.push(QueueEntry {
                node: edge.to,
                distance: candidate,
            });
        }
    }

    if distances[goal] == u64::MAX {
        return Ok(DijkstraResult {
            found: false,
            cost: None,
            path: Vec::new(),
        });
    }

    let mut path = vec![goal];
    while path.last() != Some(&start) {
        path.push(parents[path.last().unwrap()]);
    }
    path.reverse();
    Ok(DijkstraResult {
        found: true,
        cost: Some(distances[goal]),
        path,
    })
}

fn main() {
    let graph = HashMap::from([
        ("A", vec![Edge { to: "B", weight: 2 }, Edge { to: "C", weight: 7 }]),
        ("B", vec![Edge { to: "C", weight: 1 }, Edge { to: "D", weight: 5 }]),
        ("C", vec![Edge { to: "D", weight: 1 }]),
        ("D", vec![]),
    ]);
    let result = dijkstra(&graph, "A", "D").expect("valid graph");
    assert_eq!(
        result,
        DijkstraResult {
            found: true,
            cost: Some(4),
            path: vec!["A", "B", "C", "D"],
        }
    );
    println!("rust dijkstra: ok");
}
